### PR TITLE
[MAINTENANCE] Change recursion to iteration for function in parameter_container.py

### PR DESCRIPTION
**Pull request by billdirks**
This removes recursion, reduces the line count, improves performance (from ~1min to 40s on James's test set), and, in my opinion, improves readability.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

**Review by Kilo59 — APPROVED** [review]
🚀

File: great_expectations/rule_based_profiler/parameter_container.py [hunk 1]
```diff
@@ -359,31 +359,23 @@ def _build_parameter_node_tree_for_one_parameter(
...
-    else:
-        # If the fully-qualified parameter name (or "name space") is trivial (i.e., at "leaf node" / last part), then
-        # store the supplied attribute value into the given ParameterNode using leaf "parameter_name_part" name as key.
-        parameter_node[parameter_name_part] = convert_dictionary_to_parameter_node(
-            source=parameter_value
-        )
+    node: ParameterNode = parameter_node
+    for parameter_name in parameter_name_as_list[:-1]:
+        # This conditional is functionally equivalent to node.setdefault(parameter_name, ParameterNode({})).
+        # However, setdefault always evaluates its second argument which is much slower in this hot code path.
```
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/parameter_container.py]: ❤️

**Review by Shinnnyshinshin — APPROVED** [review]
Love it. Thank you @billdirks

File: great_expectations/rule_based_profiler/parameter_container.py [hunk 2]
```diff
@@ -359,31 +359,23 @@ def _build_parameter_node_tree_for_one_parameter(
...
-            parameter_node[parameter_name_part] = ParameterNode({})
-
-        _build_parameter_node_tree_for_one_parameter(
-            parameter_node=parameter_node[parameter_name_part],
-            parameter_name_as_list=parameter_name_as_list[1:],
-            parameter_value=parameter_value,
-        )
-    else:
-        # If the fully-qualified parameter name (or "name space") is trivial (i.e., at "leaf node" / last part), then
-        # store the supplied attribute value into the given ParameterNode using leaf "parameter_name_part" name as key.
```
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/parameter_container.py:383]: Would you mind adding this comment back where appropriate? I believe it now applies at line 376? At least I'm sure most of the comment applies.

**Review by NathanFarmer — COMMENTED** [review]
Just one question for my understanding.

File: great_expectations/rule_based_profiler/parameter_container.py (hunk 2 above)
Comment (billdirks) [reply great_expectations/rule_based_profiler/parameter_container.py:383]: I've added this to the docstring. I've also updated the docstring to be consistent with the google style guide. Let me know if it is still unclear.

Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/parameter_container.py:383]: Perfect! ✅

**Pull request merged**